src/h3/mod.rs

```diff
@@ -26,4 +26,778 @@
+    qpack_decoder: qpack::Decoder,
+
+    local_qpack_streams: QpackStreams,
+    peer_qpack_streams: QpackStreams,
+}
+
+/// Creates a new client-side HTTP/3 connection.
+pub fn connect(
+    quic_conn: &mut super::Connection, config: &Config,
+) -> Result<Connection> {
+    let mut http3_conn = Connection::new(config, false)?;
+
+    http3_conn.send_settings(quic_conn)?;
+    http3_conn.open_qpack_streams(quic_conn)?;
+
+    Ok(http3_conn)
+}
+
+/// Creates a new server-side HTTP/3 connection.
+pub fn accept(
+    quic_conn: &mut super::Connection, config: &Config,
+) -> Result<Connection> {
+    let mut http3_conn = Connection::new(config, true)?;
+
+    http3_conn.send_settings(quic_conn)?;
+    http3_conn.open_qpack_streams(quic_conn)?;
+
+    Ok(http3_conn)
+}
+
+impl Connection {
+    fn new(config: &Config, is_server: bool) -> Result<Connection> {
+        let initial_uni_stream_id = if is_server { 0x3 } else { 0x2 };
+
+        Ok(Connection {
+            is_server,
+
+            highest_request_stream_id: 0,
+            highest_uni_stream_id: initial_uni_stream_id,
+
+            streams: StreamMap::new(),
+
+            local_settings: ConnectionSettings {
+                num_placeholders: Some(config.num_placeholders),
+                max_header_list_size: Some(config.max_header_list_size),
+                qpack_max_table_capacity: Some(config.qpack_max_table_capacity),
+                qpack_blocked_streams: Some(config.qpack_blocked_streams),
+            },
+
+            peer_settings: ConnectionSettings {
+                num_placeholders: None,
+                max_header_list_size: None,
+                qpack_max_table_capacity: None,
+                qpack_blocked_streams: None,
+            },
+
+            control_stream_id: None,
+            peer_control_stream_id: None,
+
+            qpack_encoder: qpack::Encoder::new(),
+            qpack_decoder: qpack::Decoder::new(),
+
+            local_qpack_streams: QpackStreams {
+                encoder_stream_id: None,
+                decoder_stream_id: None,
+            },
+            peer_qpack_streams: QpackStreams {
+                encoder_stream_id: None,
+                decoder_stream_id: None,
+            },
+        })
+    }
+
+    /// Test if all local and peer Control and QPACK streams are open
+    pub fn is_established(&self) -> bool {
+        self.control_stream_id.is_some() &&
+            self.local_qpack_streams.encoder_stream_id.is_some() &&
+            self.local_qpack_streams.decoder_stream_id.is_some() &&
+            self.peer_control_stream_id.is_some() &&
+            self.peer_qpack_streams.decoder_stream_id.is_some() &&
+            self.peer_qpack_streams.encoder_stream_id.is_some()
+    }
+
+    /// Prepare a request in HTTP/3 wire format, allocate a stream ID and send
+    /// it.
+    pub fn send_request_headers(
+        &mut self, quic_conn: &mut super::Connection, headers: &[qpack::Header],
+        has_body: bool,
```

Comment:
The `has_body` flag should probably be renamed to `fin`, so you can pass it directly to `stream_send()`.